pkg/replica/hash.go

```diff
@@ -0,0 +1,400 @@
+			State: ProgressStateInProgress,
+		},
+	}
+}
+
+func (t *SnapshotHashJob) LockFile() (fileLock *flock.Flock, err error) {
+	defer func() {
+		if err != nil && fileLock != nil && fileLock.Path() != "" {
+			if err := os.RemoveAll(fileLock.Path()); err != nil {
+				logrus.Warnf("failed to remove lock file %v since %v", fileLock.Path(), err)
+			}
+		}
+	}()
+
+	err = os.MkdirAll(FileLockDirectory, 0755)
+	if err != nil {
+		return nil, err
+	}
+
+	fileLock = flock.New(filepath.Join(FileLockDirectory, HashLockFileName))
+
+	// Blocking lock
+	err = fileLock.Lock()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to fetch the file lock for hashing snapshot %v", t.SnapshotName)
+	}
+
+	return fileLock, nil
+}
+
+func (t *SnapshotHashJob) UnlockFile(fileLock *flock.Flock) {
+	fileLock.Unlock()
+}
+
+func (t *SnapshotHashJob) Execute() {
+	var err error
+	var checksum string
+
+	defer func() {
+		t.StatusLock.Lock()
+		defer t.StatusLock.Unlock()
+		t.Checksum = checksum
+		if err != nil {
+			logrus.Errorf("failed to hash snapshot %v since %v", t.SnapshotName, err)
+			t.State = ProgressStateError
+			t.Error = err.Error()
+		} else {
+			t.State = ProgressStateComplete
+		}
+	}()
+
+	// Each node can have only one snapshot hashing task at the same time per node.
+	// When the snapshot hashing task is started, the task tries to fetch the file lock
+	// (${fileLockDirectory}/hash). If the lock file is held by another task, it will stuck
+	// here and wait for the lock. The file is unlocked after the task is completed.
+	fileLock, err := t.LockFile()
+	if err != nil {
+		return
+	}
+	defer t.UnlockFile(fileLock)
+
+	modTime, err := GetSnapshotModTime(t.SnapshotName)
+	if err != nil {
+		return
+	}
+
+	// If the silent corruption is detected, don't need to recalculate the checksum.
+	// Just set SilentlyCorrupted to ture and return it.
+	isSilentlyCorrupted, err := t.isSilentCorruptionAlreadyDetected(modTime)
+	if err != nil {
+		return
+	} else if isSilentlyCorrupted {
+		t.StatusLock.Lock()
+		t.SilentlyCorrupted = true
+		t.StatusLock.Unlock()
+		return
+	}
+
+	requireRehash := true
+	if !t.Rehash {
+		requireRehash, checksum, err = t.isRehashRequired(modTime)
+		if err != nil {
+			return
+		}
+		if !requireRehash {
+			return
+		}
+	}
+
+	logrus.Infof("Starting hashing snapshot %v", t.SnapshotName)
+
+	lastHashedAt := time.Now().UTC().Format(time.RFC3339)
+	checksum, err = hashSnapshot(t.Ctx, t.SnapshotName)
+	if err != nil {
+		return
+	}
+
+	// If the silent corruption is detected, the xattr will not be overrode.
+	// The scene will be preserved.
+	if t.isSnapshotSilentlyCorrupted(checksum) {
+		t.StatusLock.Lock()
+		t.SilentlyCorrupted = true
+		t.StatusLock.Unlock()
+		// Do the best to set SilentlyCorrupted to true.
+		// Never mind if failed to set it. The checksum will be recalculated and get the SilentlyCorrupted again.
+		if err := SetSilentlyCorruptedToXattr(t.SnapshotName); err != nil {
```

Comment:
NIT: can we use a single `SetSnapshotHashInfoToXattr` instead?

Reply:
Yeah. But because I don't want to change other values in the xattr, so the `GetSnapshotHashFromToXattr` and then `SetSnapshotHashInfoToXattr` are necessary. The silent corruption is not common, the two operations will not introduce negative imact.

Reply:
Just revisit the logic. Why not directly use a defer function to blindly update the xattr? It will be updated when the modTime remains while there is a difference from the current value.

Reply:
Refactored this part as you mentioned. Thank you.